### ISSUE TITLE
Extract a utility method for applying a series of operations' unitaries

### ISSUE THREAD

**Issue by Strilanc** [quantumlib/Cirq#1394]
There are several places in the code base that loop over a series of operations while calling `cirq.apply_unitary` and managing the workspace-vs-state buffers. One in Circuit, one in sparse_simulator, and soon to be on in ParallelGateOperation.

It's relatively easy to get this code wrong, so we should have one common place that handles it consistently. Basically I'm picturing a method with this signature:

```python
def apply_unitaries(
    operations: OP_TREE,
    # defaults to enumerate(sorted(set(q for op in ops for q in op.qubits)))
    qubit_axis_map: Dict[QubitId, int] = None,
    # defaults to the zero computational basis state sized using max index in qubit axis map
    args: ApplyUnitaryArgs = None,
    ) -> np.ndarray:
    pass
```

though I'm very open to suggestions on how it could be designed better. E.g. maybe an `apply` method on `ApplyUnitaryArgs` that does the right thing somehow.

**Comment by Ashalynd** [comment]
The piece of code in density_matrix_simulator is also conceptually similar, except that it applies a channel i/o applying unitary. Is the utility method expected to cover apply_channel too?

**Comment by c-poole** [comment]
Can this be closed now that #1650 was merged or is there more to be done @Strilanc?

**Comment by Strilanc** [comment]
Yup, it can be closed!
